python/pyarrow/dataset.py

```diff
@@ -155,19 +155,16 @@ def partitioning(schema=None, field_names=None, flavor=None):
 
 def _ensure_fs(filesystem, path):
     # Validate or infer the filesystem from the path
-    from pyarrow.fs import (
-        FileSystem, LocalFileSystem, FileType, _normalize_path)
+    from pyarrow.fs import FileSystem, LocalFileSystem, _normalize_path
 
     if filesystem is None:
-        # first check if the file exists as a local (relative) file path
-        filesystem = LocalFileSystem()
         try:
-            infos = filesystem.get_file_info([path])[0]
-        except OSError:
-            return FileSystem.from_uri(path)
-
-        if infos.type == FileType.NotFound:
-            return FileSystem.from_uri(path)
+            scheme, _ = path.split('://', 1)  # noqa
```

Comment:
This is definitely wrong. URI parsing is more complex than this...

Reply:
Well, FileSystemFromUriOrPath is used by the dataset implementation, and it raises FileNotFoundError only for `file://relate-path` uris.

Reply:
Otherwise it raises here https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/uri.cc#L207

Reply:
@pitrou any suggestions how to raise FileNotFoundError properly?

Reply:
I'll take a look.

Reply:
Thanks!